pkg/upgrade/upgrades/update_invalid_column_ids_in_sequence_back_references.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package upgrades
+
+import (
+	"context"
+	"reflect"
+	"sort"
+
+	"github.com/cockroachdb/cockroach/pkg/clusterversion"
+	"github.com/cockroachdb/cockroach/pkg/jobs"
+	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descs"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlutil"
+	"github.com/cockroachdb/cockroach/pkg/upgrade"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/errors"
+)
+
+func updateInvalidColumnIDsInSequenceBackReferences(
+	ctx context.Context, _ clusterversion.ClusterVersion, d upgrade.TenantDeps, _ *jobs.Job,
+) error {
+	// Scan all sequences and repair those with 0-valued column IDs in their back references,
+	// one transaction for each repair.
+	var seqID descpb.ID
+	var done bool
```

Comment:
unfortunately I think this is hazardous. Consider a case where you retry, you might have assigned to `seqID`.  Try something like this:

```go
  var lastSeqID descpb.ID
	for {
	  var curSeqID descpb.ID
    var done bool
		if err := d.CollectionFactory.TxnWithExecutor(ctx, d.DB, d.SessionData, func(
			ctx context.Context, txn *kv.Txn, descriptors *descs.Collection, ie sqlutil.InternalExecutor,
		) (err error) {
			for {
					done, curSeqID, err = findNextTableToUpgrade(ctx, ie, lastSeqID,
					func(table *descpb.TableDescriptor) bool {
						return table.IsSequence()
					})
				if err != nil || done {
					return err
				}

				// Sequence `nextIdToUpgrade` might contain back reference with invalid column IDs. If so, we need to
				// update them with valid column IDs.
				hasUpgrade, err := maybeUpdateInvalidColumnIdsInSequenceBackReferences(ctx, txn, curSeqID, descriptors)
				if err != nil {
					return err
				}
				if hasUpgrade {
					return nil
				}
			}
		}); err != nil {
			return err
		}

		// Break out of the loop if we upgraded all sequences.
		if done {
			break
		}
		lastSeqID = curSeqID
	}
	```

Reply:
ahh, retries, good point! It should work if we add one line to your snippet

```go
  var lastSeqID descpb.ID
	for {
	  var curSeqID descpb.ID
    var done bool
		if err := d.CollectionFactory.TxnWithExecutor(ctx, d.DB, d.SessionData, func(
			ctx context.Context, txn *kv.Txn, descriptors *descs.Collection, ie sqlutil.InternalExecutor,
		) (err error) {
			for {
					done, curSeqID, err = findNextTableToUpgrade(ctx, ie, lastSeqID,
					func(table *descpb.TableDescriptor) bool {
						return table.IsSequence()
					})
				if err != nil || done {
					return err
				}

				// Sequence `nextIdToUpgrade` might contain back reference with invalid column IDs. If so, we need to
				// update them with valid column IDs.
				hasUpgrade, err := maybeUpdateInvalidColumnIdsInSequenceBackReferences(ctx, txn, curSeqID, descriptors)
				if err != nil {
					return err
				}
				if hasUpgrade {
					return nil
				}
+              lastSeqID = curSeqID
			}
		}); err != nil {
			return err
		}

		// Break out of the loop if we upgraded all sequences.
		if done {
			break
		}
		lastSeqID = curSeqID
	}
```

Otherwise it will hang

Reply:
that also does not work. Instead, I think you need to initialize `curSeqID` to `lastSeqID`above the inner for loop and use `curSeqID` to kick off the iteration.

Reply:
yeah, that's right. How about my latest change?

```go
{
	// Scan all sequences and repair those with 0-valued column IDs in their back references,
	// one transaction for each repair.
	var lastSeqID descpb.ID

	for {
		var currSeqID descpb.ID
		var done bool
		if err := d.CollectionFactory.TxnWithExecutor(ctx, d.DB, d.SessionData, func(
			ctx context.Context, txn *kv.Txn, descriptors *descs.Collection, ie sqlutil.InternalExecutor,
		) (err error) {
			currSeqID = lastSeqID
			for {
				done, currSeqID, err = findNextTableToUpgrade(ctx, ie, currSeqID,
					func(table *descpb.TableDescriptor) bool {
						return table.IsSequence()
					})
				if err != nil || done {
					return err
				}

				// Sequence `nextIdToUpgrade` might contain back reference with invalid column IDs. If so, we need to
				// update them with valid column IDs.
				hasUpgrade, err := maybeUpdateInvalidColumnIdsInSequenceBackReferences(ctx, txn, currSeqID, descriptors)
				if err != nil {
					return err
				}
				if hasUpgrade {
					return nil
				}
			}
		}); err != nil {
			return err
		}

		// Break out of the loop if we upgraded all sequences.
		if done {
			break
		}
		lastSeqID = currSeqID
	}

	return nil
}
```